### PR TITLE
Issue 886: fix(dswx-s1): apply grace period for min-bursts scenario. revise logging

### PR DESCRIPTION
**Pull request by chrisjrd**
Refs #886

## Purpose
apply grace period for min-bursts scenario. revise logging.
## Proposed Changes
- [CHANGE] logging. and apply grace period for min-bursts scenario.

## Issues
- issue-886
## Testing
- tested locally.

### CODE REVIEW

**Comment by riverma** [comment]
@chrisjrd - can you provide a reproducible command to execute your `test_evaluator.py`? I'm getting module import errors using PyTest.

**Comment by hhlee445** [comment]
Tested on **mozart** under _/export/home/hysdsops/mozart/ops/opera-pcm/tests/unit/data_subscriber/rtc_ directory. 


================================================================================ test session starts =================================================================================
platform linux -- Python 3.9.18, pytest-7.2.1, pluggy-1.0.0
rootdir: /export/home/hysdsops/mozart/ops/opera-pcm, configfile: pytest.ini
plugins: mock-3.14.0, cov-4.0.0, asyncio-0.20.3
asyncio: mode=strict
collected 8 items                                                                                                                                                                    

test_evaluator.py::test_coverage_target__when_burst_covered__and_out_of_grace_period__then_kept_in_evaluator_results[0-expected_sets0] 
----------------------------------------------------------------------------------- live log setup -----------------------------------------------------------------------------------
INFO:     600 data_subscriber.rtc.mgrs_bursts_collection_db_client:mgrs_bursts_collection_db_client.py:cached_load_mgrs_burst_db:37 - Cache loading MGRS burst database. filter_land=True
INFO:     601 data_subscriber.rtc.mgrs_bursts_collection_db_client:mgrs_bursts_collection_db_client.py:load_mgrs_burst_db:43 - Loading MGRS burst database. filter_land=True
INFO:    7024 data_subscriber.rtc.mgrs_bursts_collection_db_client:mgrs_bursts_collection_db_client.py:load_mgrs_burst_db_raw:73 - len(vector_gdf)=31500
INFO:    7051 data_subscriber.rtc.mgrs_bursts_collection_db_client:mgrs_bursts_collection_db_client.py:load_mgrs_burst_db_raw:76 - len(vector_gdf)=14687
----------------------------------------------------------------------------------- live log call ------------------------------------------------------------------------------------
INFO:    8549 data_subscriber.rtc.evaluator:evaluator.py:main:35 - coverage_target=0, min_num_bursts=None
INFO:    8549 data_subscriber.rtc.evaluator:evaluator.py:main:67 - Found len(unsubmitted_docs)=1
INFO:    8549 data_subscriber.rtc.evaluator:evaluator.py:main:74 - Found len(submitted_but_incomplete_docs)=1
INFO:    8557 data_subscriber.rtc.evaluator:evaluator.py:evaluate_rtc_products:200 - len(mbc_filtered_gdf)=87
INFO:    8558 data_subscriber.rtc.evaluator:evaluator.py:evaluate_rtc_products:217 - grouping by sliding time windows
INFO:    8591 data_subscriber.rtc.evaluator_core:evaluator_core.py:process:64 - BEGIN
INFO:    8616 data_subscriber.rtc.evaluator_core:evaluator_core.py:_find_set_coverage_in_orbit:112 - Processing orbit=20
INFO:    8633 data_subscriber.rtc.evaluator_core:evaluator_core.py:process:67 - DONE
INFO:    8633 data_subscriber.rtc.evaluator_core:evaluator_core.py:process:69 - Cleaning up the sets
INFO:    8634 data_subscriber.rtc.evaluator_core:evaluator_core.py:process:71 - Collecting as set of sets
INFO:    8634 data_subscriber.rtc.evaluator:evaluator.py:main:108 - coverage_results_short={0: ['MS_20_29', 'MS_20_30']}
INFO:    8634 data_subscriber.rtc.evaluator:evaluator.py:main:110 - Converting coverage results to evaluator results
INFO:    8634 data_subscriber.rtc.evaluator:evaluator.py:main:116 - Results that meet the target coverage (coverage_target=0), coverage_group=0: ['MS_20_29', 'MS_20_30']
INFO:    8642 data_subscriber.rtc.evaluator:evaluator.py:main:150 - mgrs_set_id='MS_20_29', number_of_bursts=1, min_num_bursts=0
INFO:    8642 data_subscriber.rtc.evaluator:evaluator.py:main:169 - Target covered burst set aged out of grace period (grace_period_minutes_remaining=datetime.timedelta(days=-1, seconds=86399, microseconds=284575)). Will process at this time. mgrs_set_id='MS_20_29', i=0
INFO:    8642 data_subscriber.rtc.evaluator:evaluator.py:main:150 - mgrs_set_id='MS_20_30', number_of_bursts=1, min_num_bursts=0
INFO:    8642 data_subscriber.rtc.evaluator:evaluator.py:main:169 - Target covered burst set aged out of grace period (grace_period_minutes_remaining=datetime.timedelta(days=-1, seconds=86399, microseconds=284208)). Will process at this time. mgrs_set_id='MS_20_30', i=0
PASSED                                                                                                                                                                         [ 12%]
test_evaluator.py::test_coverage_target__when_burst_covered__but_still_in_grace_period__then_removed_from_evaluator_results[0-expected_sets0] 
----------------------------------------------------------------------------------- live log call ------------------------------------------------------------------------------------
INFO:    8651 data_subscriber.rtc.evaluator:evaluator.py:main:35 - coverage_target=0, min_num_bursts=None
INFO:    8651 data_subscriber.rtc.evaluator:evaluator.py:main:67 - Found len(unsubmitted_docs)=1
INFO:    8651 data_subscriber.rtc.evaluator:evaluator.py:main:74 - Found len(submitted_but_incomplete_docs)=1
INFO:    8657 data_subscriber.rtc.evaluator:evaluator.py:evaluate_rtc_products:200 - len(mbc_filtered_gdf)=87
INFO:    8658 data_subscriber.rtc.evaluator:evaluator.py:evaluate_rtc_products:217 - grouping by sliding time windows
INFO:    8687 data_subscriber.rtc.evaluator_core:evaluator_core.py:process:64 - BEGIN
INFO:    8714 data_subscriber.rtc.evaluator_core:evaluator_core.py:_find_set_coverage_in_orbit:112 - Processing orbit=20
INFO:    8731 data_subscriber.rtc.evaluator_core:evaluator_core.py:process:67 - DONE
INFO:    8731 data_subscriber.rtc.evaluator_core:evaluator_core.py:process:69 - Cleaning up the sets
INFO:    8731 data_subscriber.rtc.evaluator_core:evaluator_core.py:process:71 - Collecting as set of sets
INFO:    8732 data_subscriber.rtc.evaluator:evaluator.py:main:108 - coverage_results_short={0: ['MS_20_30', 'MS_20_29']}
INFO:    8732 data_subscriber.rtc.evaluator:evaluator.py:main:110 - Converting coverage results to evaluator results
INFO:    8732 data_subscriber.rtc.evaluator:evaluator.py:main:116 - Results that meet the target coverage (coverage_target=0), coverage_group=0: ['MS_20_30', 'MS_20_29']
INFO:    8740 data_subscriber.rtc.evaluator:evaluator.py:main:150 - mgrs_set_id='MS_20_30', number_of_bursts=1, min_num_bursts=0
INFO:    8740 data_subscriber.rtc.evaluator:evaluator.py:main:165 - Target covered burst still within grace period (grace_period_minutes_remaining=datetime.timedelta(seconds=43199, microseconds=186713)). Will not process at this time. mgrs_set_id='MS_20_30', i=0
INFO:    8740 data_subscriber.rtc.evaluator:evaluator.py:main:173 - Burst set mgrs_set_id='MS_20_30' flagged for skipping further processing at this time. i=0
INFO:    8740 data_subscriber.rtc.evaluator:evaluator.py:main:150 - mgrs_set_id='MS_20_29', number_of_bursts=1, min_num_bursts=0
INFO:    8740 data_subscriber.rtc.evaluator:evaluator.py:main:165 - Target covered burst still within grace period (grace_period_minutes_remaining=datetime.timedelta(seconds=43199, microseconds=186239)). Will not process at this time. mgrs_set_id='MS_20_29', i=0
INFO:    8741 data_subscriber.rtc.evaluator:evaluator.py:main:173 - Burst set mgrs_set_id='MS_20_29' flagged for skipping further processing at this time. i=0
PASSED                                                                                                                                                                         [ 25%]
test_evaluator.py::test_native_id__when_coverage_target_A__and_bursts_found_B[100-expected_sets0] 
----------------------------------------------------------------------------------- live log call ------------------------------------------------------------------------------------
INFO:    8743 data_subscriber.rtc.evaluator:evaluator.py:main:35 - coverage_target=100, min_num_bursts=None
INFO:    8744 data_subscriber.rtc.evaluator:evaluator.py:main:50 - Supplied mgrs_set_id_acquisition_ts_cycle_indexes={'dummy'}. Adding criteria to query
INFO:    8750 data_subscriber.rtc.evaluator:evaluator.py:evaluate_rtc_products:200 - len(mbc_filtered_gdf)=87
INFO:    8751 data_subscriber.rtc.evaluator:evaluator.py:evaluate_rtc_products:217 - grouping by sliding time windows
INFO:    8786 data_subscriber.rtc.evaluator_core:evaluator_core.py:process:64 - BEGIN
INFO:    8816 data_subscriber.rtc.evaluator_core:evaluator_core.py:_find_set_coverage_in_orbit:112 - Processing orbit=20
INFO:    8840 data_subscriber.rtc.evaluator_core:evaluator_core.py:process:67 - DONE
INFO:    8840 data_subscriber.rtc.evaluator_core:evaluator_core.py:process:69 - Cleaning up the sets
INFO:    8840 data_subscriber.rtc.evaluator_core:evaluator_core.py:process:71 - Collecting as set of sets
INFO:    8841 data_subscriber.rtc.evaluator:evaluator.py:main:108 - coverage_results_short={-1: ['MS_20_30', 'MS_20_29']}
INFO:    8841 data_subscriber.rtc.evaluator:evaluator.py:main:110 - Converting coverage results to evaluator results
INFO:    8841 data_subscriber.rtc.evaluator:evaluator.py:main:114 - Skipping results that don't meet the target coverage (coverage_target=100), coverage_group=-1: ['MS_20_30', 'MS_20_29']
INFO:    8841 data_subscriber.rtc.evaluator:evaluator.py:main:134 - ignoring grace period
PASSED                                                                                                                                                                         [ 37%]
test_evaluator.py::test_native_id__when_coverage_target_A__and_bursts_found_B[1-expected_sets1] 
----------------------------------------------------------------------------------- live log call ------------------------------------------------------------------------------------
INFO:    8845 data_subscriber.rtc.evaluator:evaluator.py:main:35 - coverage_target=1, min_num_bursts=None
INFO:    8845 data_subscriber.rtc.evaluator:evaluator.py:main:50 - Supplied mgrs_set_id_acquisition_ts_cycle_indexes={'dummy'}. Adding criteria to query
INFO:    8853 data_subscriber.rtc.evaluator:evaluator.py:evaluate_rtc_products:200 - len(mbc_filtered_gdf)=87
INFO:    8854 data_subscriber.rtc.evaluator:evaluator.py:evaluate_rtc_products:217 - grouping by sliding time windows
INFO:    8890 data_subscriber.rtc.evaluator_core:evaluator_core.py:process:64 - BEGIN
INFO:    8919 data_subscriber.rtc.evaluator_core:evaluator_core.py:_find_set_coverage_in_orbit:112 - Processing orbit=20
INFO:    8936 data_subscriber.rtc.evaluator_core:evaluator_core.py:process:67 - DONE
INFO:    8937 data_subscriber.rtc.evaluator_core:evaluator_core.py:process:69 - Cleaning up the sets
INFO:    8937 data_subscriber.rtc.evaluator_core:evaluator_core.py:process:71 - Collecting as set of sets
INFO:    8937 data_subscriber.rtc.evaluator:evaluator.py:main:108 - coverage_results_short={1: ['MS_20_29', 'MS_20_30']}
INFO:    8937 data_subscriber.rtc.evaluator:evaluator.py:main:110 - Converting coverage results to evaluator results
INFO:    8937 data_subscriber.rtc.evaluator:evaluator.py:main:116 - Results that meet the target coverage (coverage_target=1), coverage_group=1: ['MS_20_29', 'MS_20_30']
INFO:    8945 data_subscriber.rtc.evaluator:evaluator.py:main:134 - ignoring grace period
PASSED                                                                                                                                                                         [ 50%]
test_evaluator.py::test_native_id__when_coverage_target_A__and_bursts_found_B[0-expected_sets2] 
----------------------------------------------------------------------------------- live log call ------------------------------------------------------------------------------------
INFO:    8948 data_subscriber.rtc.evaluator:evaluator.py:main:35 - coverage_target=0, min_num_bursts=None
INFO:    8948 data_subscriber.rtc.evaluator:evaluator.py:main:50 - Supplied mgrs_set_id_acquisition_ts_cycle_indexes={'dummy'}. Adding criteria to query
INFO:    8955 data_subscriber.rtc.evaluator:evaluator.py:evaluate_rtc_products:200 - len(mbc_filtered_gdf)=87
INFO:    8956 data_subscriber.rtc.evaluator:evaluator.py:evaluate_rtc_products:217 - grouping by sliding time windows
INFO:    8985 data_subscriber.rtc.evaluator_core:evaluator_core.py:process:64 - BEGIN
INFO:    9010 data_subscriber.rtc.evaluator_core:evaluator_core.py:_find_set_coverage_in_orbit:112 - Processing orbit=20
INFO:    9027 data_subscriber.rtc.evaluator_core:evaluator_core.py:process:67 - DONE
INFO:    9028 data_subscriber.rtc.evaluator_core:evaluator_core.py:process:69 - Cleaning up the sets
INFO:    9028 data_subscriber.rtc.evaluator_core:evaluator_core.py:process:71 - Collecting as set of sets
INFO:    9028 data_subscriber.rtc.evaluator:evaluator.py:main:108 - coverage_results_short={0: ['MS_20_30', 'MS_20_29']}
INFO:    9028 data_subscriber.rtc.evaluator:evaluator.py:main:110 - Converting coverage results to evaluator results
INFO:    9028 data_subscriber.rtc.evaluator:evaluator.py:main:116 - Results that meet the target coverage (coverage_target=0), coverage_group=0: ['MS_20_30', 'MS_20_29']
INFO:    9036 data_subscriber.rtc.evaluator:evaluator.py:main:134 - ignoring grace period
PASSED                                                                                                                                                                         [ 62%]
test_evaluator.py::test_native_id__when_min_bursts_A__and__bursts_found_B[2-expected_sets0] 
----------------------------------------------------------------------------------- live log call ------------------------------------------------------------------------------------
INFO:    9039 data_subscriber.rtc.evaluator:evaluator.py:main:35 - coverage_target=None, min_num_bursts=2
INFO:    9039 data_subscriber.rtc.evaluator:evaluator.py:main:67 - Found len(unsubmitted_docs)=1
INFO:    9039 data_subscriber.rtc.evaluator:evaluator.py:main:74 - Found len(submitted_but_incomplete_docs)=1
INFO:    9045 data_subscriber.rtc.evaluator:evaluator.py:evaluate_rtc_products:200 - len(mbc_filtered_gdf)=87
INFO:    9047 data_subscriber.rtc.evaluator:evaluator.py:evaluate_rtc_products:217 - grouping by sliding time windows
INFO:    9076 data_subscriber.rtc.evaluator_core:evaluator_core.py:process:64 - BEGIN
INFO:    9101 data_subscriber.rtc.evaluator_core:evaluator_core.py:_find_set_coverage_in_orbit:112 - Processing orbit=20
INFO:    9120 data_subscriber.rtc.evaluator_core:evaluator_core.py:process:67 - DONE
INFO:    9120 data_subscriber.rtc.evaluator_core:evaluator_core.py:process:69 - Cleaning up the sets
INFO:    9120 data_subscriber.rtc.evaluator_core:evaluator_core.py:process:71 - Collecting as set of sets
INFO:    9121 data_subscriber.rtc.evaluator:evaluator.py:main:108 - coverage_results_short={0: ['MS_20_30', 'MS_20_29']}
INFO:    9121 data_subscriber.rtc.evaluator:evaluator.py:main:110 - Converting coverage results to evaluator results
INFO:    9121 data_subscriber.rtc.evaluator:evaluator.py:main:116 - Results that meet the target coverage (coverage_target=0), coverage_group=0: ['MS_20_30', 'MS_20_29']
INFO:    9128 data_subscriber.rtc.evaluator:evaluator.py:main:150 - mgrs_set_id='MS_20_30', number_of_bursts=1, min_num_bursts=2
INFO:    9128 data_subscriber.rtc.evaluator:evaluator.py:main:152 - Burst set mgrs_set_id='MS_20_30' does not meet min_num_bursts=2. Found number_of_bursts=1.  Will not process at this time.
INFO:    9128 data_subscriber.rtc.evaluator:evaluator.py:main:173 - Burst set mgrs_set_id='MS_20_30' flagged for skipping further processing at this time. i=0
INFO:    9129 data_subscriber.rtc.evaluator:evaluator.py:main:150 - mgrs_set_id='MS_20_29', number_of_bursts=1, min_num_bursts=2
INFO:    9129 data_subscriber.rtc.evaluator:evaluator.py:main:152 - Burst set mgrs_set_id='MS_20_29' does not meet min_num_bursts=2. Found number_of_bursts=1.  Will not process at this time.
INFO:    9129 data_subscriber.rtc.evaluator:evaluator.py:main:173 - Burst set mgrs_set_id='MS_20_29' flagged for skipping further processing at this time. i=0
PASSED                                                                                                                                                                         [ 75%]
test_evaluator.py::test_native_id__when_min_bursts_A__and__bursts_found_B[1-expected_sets1] 
----------------------------------------------------------------------------------- live log call ------------------------------------------------------------------------------------
INFO:    9132 data_subscriber.rtc.evaluator:evaluator.py:main:35 - coverage_target=None, min_num_bursts=1
INFO:    9132 data_subscriber.rtc.evaluator:evaluator.py:main:67 - Found len(unsubmitted_docs)=1
INFO:    9132 data_subscriber.rtc.evaluator:evaluator.py:main:74 - Found len(submitted_but_incomplete_docs)=1
INFO:    9138 data_subscriber.rtc.evaluator:evaluator.py:evaluate_rtc_products:200 - len(mbc_filtered_gdf)=87
INFO:    9139 data_subscriber.rtc.evaluator:evaluator.py:evaluate_rtc_products:217 - grouping by sliding time windows
INFO:    9172 data_subscriber.rtc.evaluator_core:evaluator_core.py:process:64 - BEGIN
INFO:    9209 data_subscriber.rtc.evaluator_core:evaluator_core.py:_find_set_coverage_in_orbit:112 - Processing orbit=20
INFO:    9230 data_subscriber.rtc.evaluator_core:evaluator_core.py:process:67 - DONE
INFO:    9231 data_subscriber.rtc.evaluator_core:evaluator_core.py:process:69 - Cleaning up the sets
INFO:    9231 data_subscriber.rtc.evaluator_core:evaluator_core.py:process:71 - Collecting as set of sets
INFO:    9231 data_subscriber.rtc.evaluator:evaluator.py:main:108 - coverage_results_short={0: ['MS_20_30', 'MS_20_29']}
INFO:    9231 data_subscriber.rtc.evaluator:evaluator.py:main:110 - Converting coverage results to evaluator results
INFO:    9231 data_subscriber.rtc.evaluator:evaluator.py:main:116 - Results that meet the target coverage (coverage_target=0), coverage_group=0: ['MS_20_30', 'MS_20_29']
INFO:    9242 data_subscriber.rtc.evaluator:evaluator.py:main:150 - mgrs_set_id='MS_20_30', number_of_bursts=1, min_num_bursts=1
INFO:    9242 data_subscriber.rtc.evaluator:evaluator.py:main:169 - Target covered burst set aged out of grace period (grace_period_minutes_remaining=datetime.timedelta(days=-177, seconds=377, microseconds=684413)). Will process at this time. mgrs_set_id='MS_20_30', i=0
INFO:    9242 data_subscriber.rtc.evaluator:evaluator.py:main:150 - mgrs_set_id='MS_20_29', number_of_bursts=1, min_num_bursts=1
INFO:    9243 data_subscriber.rtc.evaluator:evaluator.py:main:169 - Target covered burst set aged out of grace period (grace_period_minutes_remaining=datetime.timedelta(days=-177, seconds=377, microseconds=683952)). Will process at this time. mgrs_set_id='MS_20_29', i=0
PASSED                                                                                                                                                                         [ 87%]
test_evaluator.py::test_native_id__when_min_bursts_A__and__bursts_found_B[0-expected_sets2] 
----------------------------------------------------------------------------------- live log call ------------------------------------------------------------------------------------
INFO:    9248 data_subscriber.rtc.evaluator:evaluator.py:main:35 - coverage_target=None, min_num_bursts=0
INFO:    9248 data_subscriber.rtc.evaluator:evaluator.py:main:67 - Found len(unsubmitted_docs)=1
INFO:    9248 data_subscriber.rtc.evaluator:evaluator.py:main:74 - Found len(submitted_but_incomplete_docs)=1
INFO:    9256 data_subscriber.rtc.evaluator:evaluator.py:evaluate_rtc_products:200 - len(mbc_filtered_gdf)=87
INFO:    9257 data_subscriber.rtc.evaluator:evaluator.py:evaluate_rtc_products:217 - grouping by sliding time windows
INFO:    9289 data_subscriber.rtc.evaluator_core:evaluator_core.py:process:64 - BEGIN
INFO:    9327 data_subscriber.rtc.evaluator_core:evaluator_core.py:_find_set_coverage_in_orbit:112 - Processing orbit=20
INFO:    9348 data_subscriber.rtc.evaluator_core:evaluator_core.py:process:67 - DONE
INFO:    9348 data_subscriber.rtc.evaluator_core:evaluator_core.py:process:69 - Cleaning up the sets
INFO:    9348 data_subscriber.rtc.evaluator_core:evaluator_core.py:process:71 - Collecting as set of sets
INFO:    9348 data_subscriber.rtc.evaluator:evaluator.py:main:108 - coverage_results_short={0: ['MS_20_29', 'MS_20_30']}
INFO:    9349 data_subscriber.rtc.evaluator:evaluator.py:main:110 - Converting coverage results to evaluator results
INFO:    9349 data_subscriber.rtc.evaluator:evaluator.py:main:116 - Results that meet the target coverage (coverage_target=0), coverage_group=0: ['MS_20_29', 'MS_20_30']
INFO:    9357 data_subscriber.rtc.evaluator:evaluator.py:main:150 - mgrs_set_id='MS_20_29', number_of_bursts=1, min_num_bursts=0
INFO:    9357 data_subscriber.rtc.evaluator:evaluator.py:main:169 - Target covered burst set aged out of grace period (grace_period_minutes_remaining=datetime.timedelta(days=-177, seconds=377, microseconds=569615)). Will process at this time. mgrs_set_id='MS_20_29', i=0
INFO:    9357 data_subscriber.rtc.evaluator:evaluator.py:main:150 - mgrs_set_id='MS_20_30', number_of_bursts=1, min_num_bursts=0
INFO:    9357 data_subscriber.rtc.evaluator:evaluator.py:main:169 - Target covered burst set aged out of grace period (grace_period_minutes_remaining=datetime.timedelta(days=-177, seconds=377, microseconds=569194)). Will process at this time. mgrs_set_id='MS_20_30', i=0
PASSED                                                                                                                                                                         [100%]

================================================================================== warnings summary ==================================================================================
tests/unit/data_subscriber/rtc/test_evaluator.py::test_coverage_target__when_burst_covered__and_out_of_grace_period__then_kept_in_evaluator_results[0-expected_sets0]
  /export/home/hysdsops/conda/lib/python3.9/site-packages/future/standard_library/__init__.py:65: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
    import imp

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
---------------------------- generated xml file: /export/home/hysdsops/mozart/ops/opera-pcm/tests/unit/data_subscriber/rtc/target/reports/junit/junit.xml ----------------------------
============================================================================ 8 passed, 1 warning in 9.17s ============================================================================

**Review by hhlee445 — APPROVED** [review]

**Comment by chrisjrd** [comment]
> @chrisjrd - can you provide a reproducible command to execute your `test_evaluator.py`? I'm getting module import errors using PyTest.

you need to set up your development environment. This is more involved than this UI can facilitate describing.

**Review by sjlewis-jpl — COMMENTED** [review]
Thank you for the quick turnaround on this fix.  Since this is my first review and I'm about to go on vacation, I'm leaving this review as "Comment".

Overall, I think more inline comments would help a lot with understanding the code.  Especially with very long functions like the `main`, it would be useful to have comments explaining the key logic, the purpose of loops & if-statements (especially the ones spanning many lines), etc.

I would like to see a docstring for the `main` function, and have it describe the purpose / functionality of the function at a high-level.  Also describe each of the inputs and outputs (including what should go into `*args` and `**kwargs`).  It looks like the `*args` and `**kwargs` are not used anywhere in the function.  Why are they there?  That should definitely be in their docstring entry.  If they don't serve a purpose in the function, they should be removed.

Also, I would expect to see  some explanation of the data structures being used in the `main` function, either in the function or module docstring.  Some things can be inferred by reading the code, but that's a huge barrier to being able to read and really understand what's going on.

Regarding the testing module:
I see two new test functions, but they both look identical.  It looks like the only difference is in the query results mock data.  Is there a reason for making 2 identical functions, rather than making `timedelta` another parameter to input?  Docstrings for the test functions would be a good place to explain why there are two.

If the reason you're making separate functions is to have more transparency in the pytest listing of test passes & failures, then I suggest putting all that logic into a helper function, and have each test function just call that helper with the parameters.  You'll have a single function doing the test logic, each test function will be a single line of code, and the pytest reports will be just as clear.

**Comment by hhlee445** [comment]
Tested with following command
python3 ~/mozart/ops/opera-pcm/data_subscriber/daac_data_subscriber.py query -c OPERA_L2_RTC-S1_V1   --job-queue=opera-job_worker-rtc_data_download  --chunk-size 1  -s 2024-06-26T12:20:15Z -e 2024-06-26T13:20:27Z

triggered 3 DSWx-S1 SCIFLO jobs (100% coverage)
8 jobs are in waiting state due do partial coverage ('MS_143_6', 'MS_143_3', 'MS_143_1', 'MS_143_20', 'MS_143_19', 'MS_142_24', 'MS_142_25', 'MS_142_179'], 100: ['MS_143_5', 'MS_143_4', 'MS_142_180'])

**Review by riverma — COMMENTED** [review]
Hey @chrisjrd @hhlee445 - thanks for working through this. I'm glad unit tests were added, so thanks for that. I had some comments on the tests:
- It's not clear how the tests are testing the grace period. Can you clarify if that is actually being tested?
- Most of the tests are testing `0` or `100` base cases. I'd like to see more testing > 0 threshold and more tests that test > 0 minimum burst that result in non-empty sets.
- The expected results for the `0` threshold or minimum burst sets state two MGRS tile set IDs should be returned. Why? Shouldn't we expect all MGRS tile sets to be returned in that case?

File: data_subscriber/rtc/evaluator.py [hunk 1]
```diff
@@ -178,6 +180,11 @@ def main(
     return evaluator_results
 
 
+def datetime_now():
+    """Calls datetime.now(). Encapsulated in a function for unit-testing purposes."""
```
[the remark this answers is not in the document]
Comment (chrisjrd) [reply data_subscriber/rtc/evaluator.py:184]: Sure. we can add more inline comments in the future.

the args/kwargs arguments are probably some artifact left from a previous revision and can be removed.

We'd need some more guidance for larger docstrings. We could add short docstrings where appropriate, but it's still subjective where and when it is preferred over letting the code describe itself.

I'll take a look at the 2 test functions. In general, I prefer single asserts or test case per test function. You observed that the test setup section contains different inputs: that's the explanation of why there are separate functions. 
In this case, docstrings would not be the appropriate place to document why there are 2. it should be apparent in the test function name, or in the params for parameterized tests. i'll take a look at if modifying the parameters makes more sense in this case, assuming it is the same scenario but a different case.

having a separate function for datetime allows us to mock in a custom clock. I don't agree that it adds cognitive load to the module if you are reading with the context of the calling code: `datetime.now()` vs `datetime_now()`  are equally comprehended. we decided not to mock the clock as the tests we added seemed adequate. So we could remove this separate function. I'll look into it.

File: tests/unit/data_subscriber/rtc/test_evaluator.py [hunk 2]
```diff
@@ -12,6 +14,50 @@ def setup_function():
...
 
 
+@pytest.mark.parametrize(
+    "test_coverage_target, expected_sets", [
+        (0, {"MS_20_29", "MS_20_30"}),
+    ]
+)
+@patch("data_subscriber.rtc.evaluator.es_conn_util")
+def test_coverage_target__when_burst_covered__and_out_of_grace_period__then_kept_in_evaluator_results(es_conn_util, test_coverage_target, expected_sets):
+    import data_subscriber.rtc.evaluator as evaluator
```
[the remark this answers is not in the document]
Comment (chrisjrd) [reply tests/unit/data_subscriber/rtc/test_evaluator.py:24]: it depends. if the import has a side-effect, it must be done inline in some cases. I can revisit this, but if we depend on some pcm-common or other module that is transitively imported, we may not be able to avoid some issues, and it would be preferred to delay the import until the test environment is fully set up by the time this function runs. It's more likely that it can be safely moved to the top in this case. I'll take a look.

File: tests/unit/data_subscriber/rtc/test_evaluator.py [hunk 3]
```diff
@@ -12,6 +14,50 @@ def setup_function():
     pass
 
 
+@pytest.mark.parametrize(
+    "test_coverage_target, expected_sets", [
+        (0, {"MS_20_29", "MS_20_30"}),
+    ]
+)
+@patch("data_subscriber.rtc.evaluator.es_conn_util")
+def test_coverage_target__when_burst_covered__and_out_of_grace_period__then_kept_in_evaluator_results(es_conn_util, test_coverage_target, expected_sets):
```
[the remark this answers is not in the document]
Comment (chrisjrd) [reply tests/unit/data_subscriber/rtc/test_evaluator.py:23]: i'll take a look at this, but it is very likely that any docstring will simply repeat what the function name describes. Maybe i can rename the function name for more clarity or describe the params, but we shouldn't make it a goal or a habit of paraphrasing too much code unless the main consumers of the code will get enough value: the code itself is the most accurate and maintainable documentation we can provide.

File: tests/unit/data_subscriber/rtc/test_evaluator.py [hunk 4]
```diff
@@ -12,6 +14,50 @@ def setup_function():
     pass
 
 
+@pytest.mark.parametrize(
```
[the remark this answers is not in the document]
Comment (chrisjrd) [reply tests/unit/data_subscriber/rtc/test_evaluator.py:17]: the coverage target is a percentage. at a higher level, it is presented as a percentage. It being a percentage is an implementation detail internal to this module. maybe we can add a docstring or something, but i wouldn't recommend going as far as refactor-renaming the usages, at this stage in development.

File: tests/unit/data_subscriber/rtc/test_evaluator.py [hunk 5]
```diff
@@ -12,6 +14,50 @@ def setup_function():
...
+    assert evaluator_results["mgrs_sets"] == expected_sets or evaluator_results["mgrs_sets"].keys() == expected_sets
+
+
+@pytest.mark.parametrize(
+    "test_coverage_target, expected_sets", [
+        (0, {"MS_20_29", "MS_20_30"}),
+    ]
+)
+@patch("data_subscriber.rtc.evaluator.es_conn_util")
+def test_coverage_target__when_burst_covered__but_still_in_grace_period__then_removed_from_evaluator_results(es_conn_util, test_coverage_target, expected_sets):
```
[the remark this answers is not in the document]
Comment (chrisjrd) [reply tests/unit/data_subscriber/rtc/test_evaluator.py:45]: i wanted to reuse the parameterized pattern. It's not necessary, and we can convert it to a non-parameterized test, but i think keeping it makes it easier to add additional cases. but i'll take another look at it.

File: CONTRIBUTING.md
[the remark this answers is not in the document]
Comment (chrisjrd) [reply CONTRIBUTING.md:1]: Yes, grace period is being tested.  grace period is read from settings.yaml internally by the code. grace period comparisons depend on the creation_timestamp of the record we are returning in the elasticsearch query mock, another implementation detail. That a result is kept in the evaluation results or not depends partially on the grace period comparison.

We could add additional cases, but it is not necessary depending on the scenario.

We don't deal with tile sets at this level of the code. But to address your question, we define a burst set of size 2 in the test. And the test accordingly checks at all (2) bursts are returned. So yes, we expect all bursts in a set to be returned in the 0 threshold case, and that is what the test does.

I think you are getting hung up on the literal values `0` and `100` or whatever, instead of thinking `burst set meets min` and `burst set doesn't meet min`, regardless of what the configured min is.

File: tests/unit/data_subscriber/rtc/test_evaluator.py [hunk 6]
```diff
@@ -12,6 +14,50 @@ def setup_function():
...
+        }
+    ]
+    dummy_mgrs_set_id_acquisition_ts_cycle_indexes = None
+    evaluator_results = evaluator.main(coverage_target=test_coverage_target, mgrs_set_id_acquisition_ts_cycle_indexes=dummy_mgrs_set_id_acquisition_ts_cycle_indexes)
+    assert evaluator_results["mgrs_sets"] == expected_sets or evaluator_results["mgrs_sets"].keys() == expected_sets
+
+
+@pytest.mark.parametrize(
+    "test_coverage_target, expected_sets", [
+        (0, {"MS_20_29", "MS_20_30"}),
```
[the remark this answers is not in the document]
Comment (chrisjrd) [reply tests/unit/data_subscriber/rtc/test_evaluator.py:41]: i'm reducing the test case down to the simplest constructs, because these are unit tests.

So I define a burst set (of size 2) and I define the cataloged bursts as well in the mock query result. With those definitions, I have enough data to exercise the different branches in the code. There is no need to define 40+ sized burst sets for unit testing purposes, and I avoid that here to keep the tests concise.

File: CONTRIBUTING.md
[the remark this answers is not in the document]
Comment (chrisjrd) [reply CONTRIBUTING.md:1]: I got the idea for CONTRIBUTING.md from a popular python library. That is the file that it puts those instructions in. I don't want to put dev env setup instructions in a TESTING.md file, and I don't want to modify the README with those instructions, since the README is empty.

I guess I could just rename the file, and we can update things from that point.
